gateway/app/main/events/fl_events.py

```diff
@@ -0,0 +1,161 @@
+        server_config = data.get(CYCLE.SERVER_CONFIG, None)  # Only one
+
+        # Create a new FL Process
+        processes.create_process(
+            model=serialized_model,
+            client_plans=serialized_client_plans,
+            client_protocols=serialized_client_protocols,
+            server_averaging_plan=serialized_avg_plan,
+            client_config=client_config,
+            server_config=server_config,
+        )
+
+    except Exception as e:  # Retrieve exception messages such as missing JSON fields.
+        response[RESPONSE_MSG.ERROR] = str(e)
+
+    return json.dumps(response)
+
+
+def authenticate(message: dict, socket) -> str:
+    """ New workers should receive a unique worker ID after authenticate on PyGrid platform.
+        Args:
+            message : Message body sended by some client.
+            socket: Socket descriptor.
+        Returns:
+            response : String response to the client
+    """
+    response = {}
+
+    # Create a new worker instance and bind it with the socket connection.
+    try:
+        # Create new worker id
+        worker_id = str(uuid.uuid4())
+
+        # Create a link between worker id and socket descriptor
+        handler.new_connection(worker_id, socket)
+
+        # Create worker instance
+        workers.create_worker(worker_id)
+
+        response[MSG_FIELD.WORKER_ID] = worker_id
+
+    except Exception as e:  # Retrieve exception messages such as missing JSON fields.
+        response[RESPONSE_MSG.ERROR] = str(e)
+
+    return json.dumps(response)
+
+
+def cycle_request(message: dict, socket) -> str:
+    """This event is where the worker is attempting to join an active federated learning cycle. 
+        Args:
+            message : Message body sended by some client.
+            socket: Socket descriptor.
+        Returns:
+            response : String response to the client
+    """
+    data = message[MSG_FIELD.DATA]
+    response = {}
+
+    try:
+        # Retrieve JSON values
+        worker_id = data.get(MSG_FIELD.WORKER_ID, None)
+        model_id = data.get(MSG_FIELD.MODEL, None)
+        version = data.get(CYCLE.VERSION, None)
+        ping = data.get(CYCLE.PING, None)
+        download = data.get(CYCLE.DOWNLOAD, None)
+        upload = data.get(CYCLE.UPLOAD, None)
+
+        # Retrieve the worker
+        worker = workers.get_worker(worker_id)
+
+        # The last time this worker was assigned for this model/version.
+        last_participation = processes.last_participation(worker_id, model_id, version)
+
+        # Retrieve the last cycle created for this model_id/version
+        cycle = processes.get_cycle(model_id, version)
+
+        _accepted = False
+
+        if worker and cycle:
+            _accepted = cycle.assign(worker_id, upload, download, last_participation)
+
+        ### MOCKUP ###
+
+        # Build response
+        if _accepted:
+            worker.register_cycle(cycle.hash, cycle)
```

Comment:
Registering a cycle in a worker in the worker after I accept it allows us to retrieve the cycle through the `worker_id`, right? Just to know, when we start storing the data in a database, will we still continue like this? Having cycles in workers and workers in cycles?

Reply:
In fact, we’re not saving workers into cycles, just their IDs.
Also, we're not creating new instances of the same cycle to store in a worker. Are just pointing them to the same instance that we saved into the Controller.
Therefore, we have only one intance of Cycle for each **model_id**, **version** and **sequence_number**. This instance is shared between all workers by pointers and they can acess properties of this instance (such as model, protocoll and plans) by different hash_codes associated with their accounts.

Reply:
To retreive the cycle pointer through the worker you'll need the `request_key`(hash_code) associated to its account.

Reply:
:+1: